File: src/lib/keyvault.ts
import { invoke } from "@tauri-apps/api/core";

/**
 * Front-facing API for the OS keychain. Accounts are host UUIDs.
 *
 * Calls are silent: errors are caught and logged so the UI can fall back to
 * prompting the user instead of breaking the flow.
 */
export const keyvaultApi = {
  async save(hostId: string, password: string): Promise<void> {
    await invoke("save_host_password", { hostId, password });
  },
  async get(hostId: string): Promise<string | null> {
    try {
      return (await invoke<string | null>("get_host_password", { hostId })) ?? null;
    } catch (e) {
      console.warn("keyvault.get:", e);
      return null;
    }
  },
  async delete(hostId: string): Promise<boolean> {
    try {
      return await invoke<boolean>("delete_host_password", { hostId });
    } catch (e) {
      console.warn("keyvault.delete:", e);
      return false;
    }
  },
  async has(hostId: string): Promise<boolean> {
    try {
      return await invoke<boolean>("has_host_password", { hostId });
    } catch (e) {
      console.warn("keyvault.has:", e);
      return false;
    }
  },
};

File: src-tauri/src/commands/keyvault.rs
//! Tauri commands for the OS keychain. Account ids in scope today are host UUIDs.

use crate::keyvault;
use crate::AppError;

#[tauri::command]
pub fn save_host_password(host_id: String, password: String) -> Result<(), AppError> {
    Ok(keyvault::set_secret(&host_id, &password)?)
}

#[tauri::command]
pub fn get_host_password(host_id: String) -> Result<Option<String>, AppError> {
    Ok(keyvault::get_secret(&host_id)?)
}

#[tauri::command]
pub fn delete_host_password(host_id: String) -> Result<bool, AppError> {
    Ok(keyvault::delete_secret(&host_id)?)
}

#[tauri::command]
pub fn has_host_password(host_id: String) -> Result<bool, AppError> {
    Ok(keyvault::get_secret(&host_id)?.is_some())
}
